libclamav/dsig.c

```diff
@@ -81,37 +81,59 @@ static char cli_ndecode(unsigned char value)
     return -1;
 }
 
-static unsigned char *cli_decodesig(const char *sig, unsigned int plen, fp_int e, fp_int n)
+static unsigned char *cli_decodesig(const char *sig, unsigned int plen, BIGNUM *e, BIGNUM *n)
 {
     int i, slen = strlen(sig), dec;
     unsigned char *plain;
-    fp_int r, p, c;
+    BIGNUM *r, *p, *c;
+    BN_CTX *bn_ctx;
 
-    fp_init(&r);
-    fp_init(&c);
+    r = BN_new();
+    p = BN_new();
+    c = BN_new();
+    bn_ctx = BN_CTX_new();
+
+    if (!r || !p || !c || !bn_ctx)
+        goto err;
+
+    BN_zero(c);
     for (i = 0; i < slen; i++) {
         if ((dec = cli_ndecode(sig[i])) < 0) {
-            return NULL;
+            goto err;
         }
-        fp_set(&r, dec);
-        fp_mul_2d(&r, 6 * i, &r);
-        fp_add(&r, &c, &c);
+        if (!BN_set_word(r, dec))
+			goto err;
+        if (!BN_lshift(r, r, 6 * i))
+			goto err;
+
+        if (!BN_add(c, c, r))
+			goto err;
     }
 
     plain = (unsigned char *)cli_calloc(plen + 1, sizeof(unsigned char));
     if (!plain) {
         cli_errmsg("cli_decodesig: Can't allocate memory for 'plain'\n");
-        return NULL;
-    }
-    fp_init(&p);
-    fp_exptmod(&c, &e, &n, &p); /* plain = cipher^e mod n */
-    fp_set(&c, 256);
-    for (i = plen - 1; i >= 0; i--) { /* reverse */
-        fp_div(&p, &c, &p, &r);
-        plain[i] = MP_GET(&r);
+        goto err;
     }
+    BN_zero(p);
+    if (!BN_mod_exp(p, c, e, n, bn_ctx))
+		goto err;
```

Comment:
indentation is really off. 

I also prefer `{ }` curly block even for 1-liner if conditions/loops whatever.  Just in case someone adds something in between like a log message or whatever they don't accidentally introduce a bug (seen it happen).

Reply:
done.